pages/builds/writing_build_scripts.md.erb

```diff
@@ -1,90 +1,145 @@
+For a full list of options, see the [Bash Reference Manual](https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html).
 
-## Debugging your environment
+<section class="Docs__troubleshooting-note">
+  <h1>Unbound variable errors</h1>
+  <p>Note that while the <code>u</code> option is generally a good default to use for all build scripts, it can cause some tools like <a href="https://rvm.io">rvm</a> to fail with “unbound variable” errors. If you encounter errors, remove the <code>u</code> from the set command in your script.</p>
+</section>
 
-The first step to debugging script is to output the environment. This will show any [environment variables set up by the Buildkite Agent](/docs/builds/environment-variables), as well as other important environment variables such as `$PATH`.
+## Capturing exit status
 
-To add the environment to your build output simply execute `env` at the start of your build script.
+Build scripts can sometimes contain commands that shouldn't affect the overall exit status. For example, take the following script:
 
-```
+```bash
 #!/bin/bash
 
-env
-```
+run_tests
+
+clean_up
+``` 
+
+Running this script will exit with the status returned by the final command, `clean_up`. However, what we really care about is the exit status of the first command, `run_tests`.
 
-You can also set the `-x` option which will print out each command, fully expanded, before executing it.
+By using a variable to store the exit status of `run_tests`, we can run additional commands (like `clean_up`) whilst still returning the original exit status. For example:
 
+```bash
+#!/bin/bash
+
+# Note the lack of 'e' option here
+set -uo pipefail
+
+# Run the command we're most interested in
+run_tests
+
+# Capture the last exit status in a variable
+TEST_STATUS=$?
+
+# Run additional commands
+clean_up
+
+# Finish with the status of the original command
+exit $TEST_STATUS
 ```
+
+Using this technique gives you complete control over the exit code of your script, and the final success or failure of your build job.
+
+## Debugging your environment
+
+The first step in debugging your build script is to view the environment variables from the Buildkite web interface:
+
+<%= image "viewing-job-environment-variables.png", width: 1044/2, height: 542/2, alt: 'Screenshot of viewing a job’s environment tab' %>
+
+There may be additional environment variables available in your build job that don’t appear in this list, such as ones set by your own [agent hooks](/docs/agent/hooks). To debug these, you can print them using `echo $SOME_VAR` before the command you're wanting to run. For example:
+
+```bash
 #!/bin/bash
 
-set -x
+echo "$PATH"
+
+some_command
 ```
 
-## Capturing Exit Status
+As a last resort, you can execute `env` to print out all the environment variable names and their values. If you use `env` you should ensure you don't leak private keys or other information, by filtering the output using a tool such as `grep` (or `egrep`).
```

Comment:
Maybe something like 'If you require more environment information' or something rather than 'as a last resort'? It sounds like they'll have no other options if they can't solve their problem after executing 'env'. I think the second sentence conveys the 'this isn't the best option' idea that I guess you were going for with 'as a last resort'?